pkg/v1/abstractions/abstractions.go

```diff
@@ -0,0 +1,65 @@
+package abstractions
+
+import (
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+)
+
+// Better names for these might be ArtifactSource or ArtifactSink.
+
+// Source represents an abstraction from which we can read artifacts.
+//
+// Descriptors() and Tags() are useful for discovering what's available from a Source.
+// Image(), Index(), and Layer() are useful for actually fetching content from that Source.
+//
+// For example:
+// * a tarball can contain multiple images, exposing its contents via Tags.
+// * a remote repository can act as a source, exposing its contents via Tags (/tags/list).
+// * any image index can act as a source, exposing its children via Descriptors.
+// * an OCI image layout acts as a source, exposing contents of index.json via Descriptors.
+type Source interface {
+	Descriptors() []v1.Descriptor
+	Tags() []name.Tag
+
+	Image(v1.Hash) v1.Image
+	Index(v1.Hash) v1.ImageIndex
+	Layer(v1.Hash) v1.Layer
+
+	// This could be a name.Tag or a v1.Hash or a name.Digest but golang doesn't have method overloading :(
+	// Not sure best way to handle it.
+	//
+	// Perhaps "ImagePointer" or something would be the union of (name.Tag + name.Digest + v1.Hash).
+	//
+	// name.Reference is almost correct, but sometimes we have a v1.Hash without a repo.
+	// We could use them for sentinel repos?
+	//
+	// The Source probably knows if the string is supposed to be a v1.Hash or a name.Reference...
+	ImageDescriptor(string) ImageDescriptor
```

Comment:
Why I feel `ImageDescriptor` is unnecessary provided you can already get image, index and layer via the `Source`?

Reply:
I imagine each package will implement `Source` as a way to expose images from a given context.

The problem is that there's two scenarios:
1. I have a descriptor, so I know exactly what I'm going to be fetching.
2. I have a reference (tag or digest), so I don't know what I'm going to be fetching.

I think we need something to represent that ambiguous-at-compile-time result, which would be `ImageDescriptor`. I think we *also* need something to represent a source of content that various packages can implement, which we can then build more sophisticated operations than `Image` and `Write`, since we'd have a little bit of a better API.

Reply:
I think it will help me understand if I know when you will call `Source.Image(v1.Hash)` and when you will call `Source.ImageDescriptor(string).Image(v1.Hash)`?

I do like the abstraction on `Source` and `Sink`. The implication seems you will have to have `struct` in each package to implement the interface (while currently you call static functions in the packages). This could be a large change.

Reply:
If a `Source` is able to populate `Descriptors`, then I can iterate over that and know what I'm pulling based on the descriptor's mediaType.

If a `Source` isn't able to poplulate `Descriptors`, then I might just be pulling *something* by digest. I'm not sure what that is, so I'd use `ImageDescriptor` to pull it and inspect what I get back.

It seems like the interface I have here for `Source` is really two abstractions glued into one because they're similar. We might want something like `IndexedSource` and `TaggedSource` and `Source` that are slightly different but all implement a simple base `Source`? Not sure.

`Source` -> can give you image, index, layer, or imagedescriptor

`IndexedSource` -> can also give you a list of Descriptors
`TaggedSource` -> can also give you a list of Tags

Reply:
Two questions.

1.
> `Source` -> can give you image, index, layer, or imagedescriptor
Not sure about how to deal with layer. But isn't `ImageDescriptor` = image + index? What's the reason to duplicate?

2.
> `IndexedSource` -> can also give you a list of Descriptors
Why not let `Source` populate a list of `ImageDescriptor`? What's the difference between getting a `Descriptor` to know the type vs. getting a `ImageDescriptor` to know the type?

Reply:
1:

Not sure about layers, but I think it would make certain things easier to implement if we had a better way to read blobs from a random source. I might be trying to solve too many problems at once, but if you had something that can produce `ImageDescriptor` and `Layer`, you can write a generic `Pull`.

2:

I think a `Descriptor` is lighter weight. It's just a pointer to a thing, whereas an `ImageDescriptor` is a partially pulled image that could be an Index or could be a Manifest. I think `ImageDescriptor` is a bad name and confuses this. Maybe something like `FetchResult` is better.